addon/devtoolsPanel.js

```diff
@@ -1,47 +1,232 @@
 }
 
-async function labelInspectedElement() {
-    /**
-     * The function embedded herein returns an "index path" to the inspected
-     * element. A return value like [1, 4, 0] means the element could be found by
-     * going into the 0th element of the document (usually an <html> tag), then the
-     * 4th child of that tag, then finding the 1st child of that tag.
-     */
-    backgroundPort.postMessage({type: 'label',
-                                tabId: browser.devtools.inspectedWindow.tabId,
-                                inspectedElement: await inspectedElementPath(),
-                                label: document.getElementById('labelField').value});
+// Returns a list of objects with the following attributes:
+//   preview: html preview of element
+//   label: the fathom label, or an empty string if unlabeled
+//   path: css path to element
+//   inspected: true if this element === $0
+function getLabeled() {
+    let evalScript = `
+        (function getLabeled() {
+            function elementMeta(el, isInspected) {
+                return {
+                    preview: el.outerHTML.replace(/^([^>]+>)[\\s\\S]*$/, '$1'),
+                    label: el.dataset.fathom || '',
+                    path: Simmer(el),
+                    inspected: isInspected,
+                };
+            }
+
+            let list = [];
+            if ($0) {
+                list.push(elementMeta($0, true));
+            }
+            for (const el of document.querySelectorAll('[data-fathom]')) {
+                if ($0 && el === $0) {
+                    continue;
+                }
+                list.push(elementMeta(el, false));
+            }
+            return list;
+        })();
+    `;
+    return resultOfEval(evalScript);
 }
-document.getElementById('labelButton').addEventListener('click', labelInspectedElement);
 
-async function freezePage() {
-    backgroundPort.postMessage({type: 'freeze',
-                                tabId: browser.devtools.inspectedWindow.tabId,
-                                inspectedElement: await inspectedElementPath(),
-                                options: {shouldScroll: false, wait: 0}});
+// Draw the table with the list of existing labeled elements.
+// If an element is inspected/selected with devtools (i.e. $0 is defined), it
+// will always be the first row in the table.
+function updateLabeledTable(labeled) {
+
+    function addTextCell(row, text, isHeader=false) {
+        const td = document.createElement(isHeader ? 'th' : 'td');
+        td.appendChild(document.createTextNode(text));
+        row.appendChild(td);
+        return td;
+    }
+
+    function inspectLabeled(event) {
+        event.preventDefault();
+        const js = 'inspect(document.querySelector("' + event.target.dataset.path + '"))';
+        browser.devtools.inspectedWindow.eval(js)
+            .catch((e) => {
+                console.error(e);
```

Comment:
There has got to be a better way of making sure uncaught exceptions get logged than doing it by hand every time, but I haven't found one yet.